### ISSUE TITLE
Issue with Auto save of ARM context

### ISSUE THREAD

**Issue by Ajay-MS** [Azure/azure-powershell#6680]
### Description

We are facing following issues with context autosave. ( Enable-AzureRmContextAutoSave)

1. We have a scenario where the same machine (agent system) runs in a different context. It means everytime we run connect-Azure-AzureRmAccount cmdlet with a different service principal. We run commands in the following order
```
Connect-AzureRMAccount -ServicePrincipal -Tenant $TenentId1 -Credential $psCredential
Set-AzureRmContext -SubscriptionId $subscriptionId1
```

Connect-AzureRmAccount is not adding new account at the top of context list saved on the machine.        Hence  Set-AzureRmContext cmdlet failed with the error that "Invalid tenet Id and subscription Id" because it checks in topmost context. 

In order to let the above scenario work, I need to run the cmdlet to clear all existing context. 
```
Clear-AzureRmContext
```

2.  If context autosave is disabled, the Connect-AzureRmAccount cmdlet is not adding service principal having permission at management group level.   On enabling context autosave, it able to add service principal successfully. 

### Module Version

Version: 6.2.1 and 6.3

### Environment Data

Name                           Value                                                                                                                                                                        
----                           -----                                                                                                                                                                        
PSVersion        -               5.1.16299.547                                                                                                                                                                
PSEdition             -         Desktop                                                                                                                                                                      
PSCompatibleVersions      -     {1.0, 2.0, 3.0, 4.0...}                                                                                                                                                      
BuildVersion       -            10.0.16299.547                                                                                                                                                               
CLRVersion          -           4.0.30319.42000                                                                                                                                                              
WSManStackVersion        -      3.0                                                                                                                                                                          
PSRemotingProtocolVersion    -  2.3                                                                                                                                                                          
SerializationVersion     -      1.1.0.1

Please let me know if any other details are being required.

**Comment by maddieclayton** [comment]
@Ajay-MS Rather than using Set-AzureRmContext, which edits the current context you are in, please use Select-AzureRmContext, which will select a different context with entirely different information.  Thus, instead of running Connect-AzureRmAccount and then Set-AzureRmContext for each account, you can instead set up all of your account manually (not in VSTS), then simply use Select-AzureRmContext to change accounts/subscriptions in the VSTS task.

Please let me know if this works for you.

**Comment by Ajay-MS** [comment]
@maddieclayton 

Thanks for your feedback. I will validate this and will update you. 

Any feedback on second scenario mentioned above ( regarding management group).

**Comment by SumiranAgg** [comment]
@maddieclayton We also observed, even after Disable-AzureRmContextAutoSave, if we try to login using different Service principal credentials having access to same subscription, previous context persists.
For ex. If I login using Service principal "a" having access to subscription "sub1", the context gets added to list only if context with "sub1" doesn't already exist.
This forces us to clear context if we wish to login using different user credentials.
We can select-azureRmContext (as per your suggestion) only if context exists in the list.

**Comment by maddieclayton** [comment]
@Ajay-MS @SumiranAgg This is an issue we fixed in 6.5.0, so please update to the latest version in order to get this fix.

@Ajay-MS Can you please update to the latest version and let us know if you are still seeing the same issue.  I have a feeling the same bug fix might take care of your issue.

**Comment by samuelbcollie** [comment]
@maddieclayton I'm still running into this issue with the latest version of Azure powershell. I've started using a workaround suggested [here](https://github.com/Microsoft/vsts-tasks/issues/7049#issuecomment-405835019) but my releases still fail more than half the time with a login error.

**Comment by almmechanics** [comment]
Same issue is with 6.6.0. The resolution i've used is to execute:

    Clear-AzureRmContext 

as a PowerShell task in VSTS *before* running any Azure PowerShell tasks. This was based on the information from [Azure Powershell task fails on AzureRM 6.4.0 when switching between tenants](https://github.com/Microsoft/vsts-tasks/issues/7689) issue.

**Comment by SumiranAgg** [comment]
@maddieclayton The issue is not resolved in version 6.5.0.
Can you try the scenario I described in previous comment?
Try to login to same subscription with different service principal credentials. I am unable to do so in PS version 6.*.

**Comment by markcowl** [comment]
@SumiranAgg @almmechanics Yes, this issue wasn't fixed in the latest version, it was fixed in a PR that was merged this week (#6797), which will be released on the 14th.  Note that this problem is irrespective of whether you use context autosave or not.

The underlying issue is that the default context name when logging in uses the subscription ID only.  As a workaround until the fix is released, if you use the ```ContextName``` parameter when logging in, this will automatically save the context under that name, and you will be able to select it.

So, for example, if I login selecting the same subscription using two different service principals, and give each login a unique context name:

```powershell
Connect-AzureRmAccount -ServicePrincipal ... -ContextName "ContextWithPrincipal1"
Connect-AzureRmAccount -ServicePrincipal ..., -ContextName "ContextWithPrincipal2"
```

I can then select either context
```powershell
Select-AzureRmContext "ContextWithPrincipal1"
... actions using service principal 1
Select-AzureRmContext "ContextWithPrincipal2"
... actions using service principal 2
```

**Comment by almmechanics** [comment]
Sadly, if this is being invoked as part of an 'Azure PowerShell' task on a VSTS hosted agent, then there is no choice in the context. 

Any previous use of the private agent will, at present, cache the endpoint using the default behaviour of Azure PowerShell.

**Comment by markcowl** [comment]
@almmechanics The default behavior in this case should be fixed in the next release. 

I am not sure I understand the scenario, though.  Are you saying that you are intentionally logging in as two different service princpals as part of the same VSTS task?  Generally you want to have a one-to-one correspondence between tasks and the principals they use.

**Comment by almmechanics** [comment]
No, my original problem was 2 Azure Powershell tasks in one build. Each targeting a unique resource group on a private build agent. 

I look forward to be revised azure powershell code module and appropriate vsts task and then all will be well.

**Comment by MiYanni** [comment]
Closing as fixed in 6.7
